### ISSUE TITLE
Crash in vendor subcommand

### ISSUE THREAD

**Issue by criemen** [bazelbuild/bazel#23521]
### Description of the bug:

```
 criemen@Corneliuss-MBP  ~/repos/semmle-code   main ±  bazel vendor --vendor_dir=vendor_src //...
Starting local Bazel server and connecting to it...
INFO: Invocation ID: a3f127d2-153b-400d-8a1e-55c75c5d6396
Loading: 1 packages loaded
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'REPOSITORY_DIRECTORY:@@WORKSPACE.bazel' (requested by nodes '[/private/var/tmp/_bazel_criemen/267ec2c545f0139b2caed9502387afe1]/[external/WORKSPACE.bazel]')
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:550)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:414)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.ClassCastException: class com.google.devtools.build.lib.packages.InputFile cannot be cast to class com.google.devtools.build.lib.packages.Rule (com.google.devtools.build.lib.packages.InputFile and com.google.devtools.build.lib.packages.Rule are in unnamed module of loader 'app')
        at com.google.devtools.build.lib.packages.Package.getRule(Package.java:615)
        at com.google.devtools.build.lib.repository.ExternalPackageHelper$ExternalPackageRuleExtractor.processAndShouldContinue(ExternalPackageHelper.java:144)
        at com.google.devtools.build.lib.repository.ExternalPackageHelper.iterateWorkspaceFragments(ExternalPackageHelper.java:118)
        at com.google.devtools.build.lib.repository.ExternalPackageHelper.getRuleByName(ExternalPackageHelper.java:52)
        at com.google.devtools.build.lib.rules.repository.RepositoryDelegatorFunction.getRepoRuleFromWorkspace(RepositoryDelegatorFunction.java:475)
        at com.google.devtools.build.lib.rules.repository.RepositoryDelegatorFunction.getRepositoryRule(RepositoryDelegatorFunction.java:360)
        at com.google.devtools.build.lib.rules.repository.RepositoryDelegatorFunction.compute(RepositoryDelegatorFunction.java:154)
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:461)
        ... 7 more
```
Presumably some symlink or something in our source root messes up the vendor subcommand - it shouldn't crash with an exception though.

### Which operating system are you running Bazel on?

MacOS

### What is the output of `bazel info release`?

development

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

5f5d70b6c4d2fb1a889479569107f1692239e8a7 via bazelisk (it's a commit from the 7.4.0 branch)

**Comment by fmeum** [comment]
@meteorcloudy

**Comment by meteorcloudy** [comment]
> `REPOSITORY_DIRECTORY:@@WORKSPACE.bazel`

This looks very weird, do you actually have a repo called `WORKSPACE.bazel` somehow? Can you provide a minimal reproducible case?

**Comment by criemen** [comment]
> do you actually have a repo called WORKSPACE.bazel somehow?

No, unless one of our transitive dependencies has one (that'd be weird though).

> Can you provide a minimal reproducible case?

Not anytime soon, but if you're fine with waiting then I'll get back to you eventually.

**Comment by moroten** [comment]
After running `bazel vendor`, I got stuck with the same error when running `bazel test` using Bazel 7.3.1:
`java.lang.RuntimeException: Unrecoverable error while evaluating node 'REPOSITORY_DIRECTORY:@@WORKSPACE' (requested by nodes '[/<redacted>/e9607a41d9]/[external/WORKSPACE]')`

**Comment by moroten** [comment]
The crash loop does not get resolved by `bazel clean --expunge`, only by removing the `vendor` directory.

I can reproduce it by running
```
touch WORKSPACE
mkdir vendor
ln -s $(bazel info output_base)/external vendor/bazel-external
bazel test ...
```

**Comment by moroten** [comment]
Duplicates #22740

**Comment by meteorcloudy** [comment]
https://github.com/bazelbuild/bazel/pull/23771 should help avoid this. A temporary solution is to add the vendor dir into `.bazelignore`.

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.4.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.4.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=7.4.0rc1`. Thanks!
